src/torch/SIRF_torch.py

```diff
@@ -0,0 +1,269 @@
+try:
+    import torch
+    import sirf
+    
+    from sirf.STIR import *
+except ModuleNotFoundError:
+    raise ModuleNotFoundError('Failed to import torch. Please install PyTorch first.')
+
+# based on 
+# https://github.com/educating-dip/pet_deep_image_prior/blob/main/src/deep_image_prior/torch_wrapper.py
+
+
+def sirf_to_torch(
+        sirf_src: object,
+        device: torch.device,
+        requires_grad: bool = False
+        ) -> torch.Tensor:
+
+    # use torch.tensor to infer data type
+    return torch.tensor(sirf_src.as_array(), requires_grad=True).to(device)
+
+def torch_to_sirf_(
+        torch_src: torch.Tensor,
+        sirf_dest: object,
+        ) -> object:
+
+    # This is an in-place operation - CAREFUL
+    # Only really to be used within torch.autograd.Function
+    return sirf_dest.fill(torch_src.detach().cpu().numpy())
+
+
+class _ObjectiveFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx,
+            torch_image: torch.Tensor,
+            sirf_image_template: object,
+            sirf_obj_func: object
+            ):
+
+        device = torch_image.device
+        sirf_image_template = torch_to_sirf_(torch_image, sirf_image_template)
+        # Negative for Gradient Descent as per torch convention
+        # HERE WE ARE COPYING THE VALUE
+        value = torch.tensor(-sirf_obj_func.get_value(sirf_image_template)).to(device)
+        if torch_image.requires_grad:
+            ctx.device = device
+            ctx.sirf_image_template = sirf_image_template
+            ctx.sirf_obj_func = sirf_obj_func
+            # ensure value is a tensor with requires_grad=True
+            return value.requires_grad_(True)
+        else:
+            return value
+
+    @staticmethod
+    def backward(ctx,
+            grad_output: torch.Tensor
+            ):
+            
+        sirf_obj = ctx.sirf_obj_func
+        sirf_image_template = ctx.sirf_image_template
+        device = ctx.device
+        # Negative for Gradient Descent as per torch convention
+        sirf_grad = -sirf_obj.get_gradient(sirf_image_template)
+        grad = sirf_to_torch(sirf_grad, device, requires_grad=True)
+        return grad_output*grad, None, None, None
+
+
+class _ObjectiveGradient(torch.autograd.Function):
```

Comment:
currently commented and not implemented, just checking my understanding is correct of this implementation: notebooks/Deep_Learning_listmode_PET/snippets/solution_4_1.py.  @KrisThielemans